src/v/storage/segment_appender.cc

```diff
@@ -374,39 +374,40 @@ ss::future<> segment_appender::do_next_adaptive_fallocation() {
 
 ss::future<> segment_appender::maybe_advance_stable_offset(
   const ss::lw_shared_ptr<inflight_write>& write) {
+    vassert(!_inflight.empty(), "expected non-empty inflight set");
+    write->set_state(write_state::DONE);
```

Comment:
Am I missing something or were we previously only setting this in the "no callbacks" branch? Was that a bug?

Reply:
@StephanDollberg - yes, but I don't think it was a bug.

Basically the "callbacks" branch is the case where at least 1 write is popped off of the `_inflight` array. This only happens if the passed-in write (`write`) is the same as the first write in the array, in which case it will be popped off of the array so it doesn't really matter what state it is in: it never gets set to DONE but it's also being deleted right now so it doesn't matter.

This change I made stops making the first write a special case, it just unconditionally sets it to DONE and then pops zero or more DONE writes from the front (so the setting is necessary).

I triggered the refactoring because I'm now tracking _inflight_dispatched so I wanted to have only 1 place where we logically "finished" the write, set the state to DONE and decremented the inflight count.

Reply:
Though ultimately this did not pay off because it turns out we need to decrement `_inflight_dispatched` at different places depending on whether there are flushes to process or not.

I still think it's a simplification tho the original impetus has evaporated.